log/lvl.go

```diff
@@ -77,108 +64,124 @@ var debugMut sync.RWMutex
-	caller := fmt.Sprintf(fmtstr, name, line)
-	if StaticMsg != "" {
-		caller += "@" + StaticMsg
-	}
-	message := fmt.Sprintln(args...)
-	bright := lvl < 0
-	lvlAbs := lvl
-	if bright {
-		lvlAbs *= -1
-	}
 
-	for _, l := range listeners {
-		// Send the listeners lines with no newlines on them.
-		l.Log(lvlAbs, message[0:len(message)-1])
-	}
-	// Now that the listeners got their copy, check level.
-	if lvl > debugVisible {
-		return
-	}
+		fmtstr := ""
+		if l.useColors {
+			// Only adjust the name and line padding if we also have color.
+			if len(name) > NamePadding && NamePadding > 0 {
+				NamePadding = len(name)
+			}
+			if len(lineStr) > LinePadding && LinePadding > 0 {
+				LinePadding = len(name)
+			}
+		}
+		fmtstr = fmt.Sprintf("%%%ds: %%%dd", NamePadding, LinePadding)
+		caller := fmt.Sprintf(fmtstr, name, line)
+		if StaticMsg != "" {
+			caller += "@" + StaticMsg
+		}
+		message := fmt.Sprintln(args...)
+		bright := lvl < 0
+		lvlAbs := lvl
+		if bright {
+			lvlAbs *= -1
+		}
 
-	lvlStr := strconv.Itoa(lvlAbs)
-	if lvl < 0 {
-		lvlStr += "!"
-	}
-	switch lvl {
-	case lvlPrint:
-		fg(ct.White, true)
-		lvlStr = "I"
-	case lvlInfo:
-		fg(ct.White, true)
-		lvlStr = "I"
-	case lvlWarning:
-		fg(ct.Green, true)
-		lvlStr = "W"
-	case lvlError:
-		fg(ct.Red, false)
-		lvlStr = "E"
-	case lvlFatal:
-		fg(ct.Red, true)
-		lvlStr = "F"
-	case lvlPanic:
-		fg(ct.Red, true)
-		lvlStr = "P"
-	default:
-		if lvl != 0 {
-			if lvlAbs <= 5 {
-				colors := []ct.Color{ct.Yellow, ct.Cyan, ct.Green, ct.Blue, ct.Cyan}
-				fg(colors[lvlAbs-1], bright)
+		lvlStr := strconv.Itoa(lvlAbs)
+		if lvl < 0 {
+			lvlStr += "!"
+		}
+		switch lvl {
+		case lvlPrint:
+			fg(l, ct.White, true)
+			lvlStr = "I"
+		case lvlInfo:
+			fg(l, ct.White, true)
+			lvlStr = "I"
+		case lvlWarning:
+			fg(l, ct.Green, true)
+			lvlStr = "W"
+		case lvlError:
+			fg(l, ct.Red, false)
+			lvlStr = "E"
+		case lvlFatal:
+			fg(l, ct.Red, true)
+			lvlStr = "F"
+		case lvlPanic:
+			fg(l, ct.Red, true)
+			lvlStr = "P"
+		default:
+			if lvl != 0 {
+				if lvlAbs <= 5 {
+					colors := []ct.Color{ct.Yellow, ct.Cyan, ct.Green, ct.Blue, ct.Cyan}
+					fg(l, colors[lvlAbs-1], bright)
+				}
 			}
 		}
-	}
-	str := fmt.Sprintf(": (%s) - %s", caller, message)
-	if showTime {
-		ti := time.Now()
-		str = fmt.Sprintf("%s.%09d%s", ti.Format("06/02/01 15:04:05"), ti.Nanosecond(), str)
-	}
-	str = fmt.Sprintf("%-2s%s", lvlStr, str)
-	if lvl < lvlInfo {
-		fmt.Fprint(stdErr, str)
-	} else {
-		fmt.Fprint(stdOut, str)
-	}
-	if useColors {
-		ct.ResetColor()
+		str := fmt.Sprintf(": (%s) - %s", caller, message)
+		if l.showTime {
+			ti := time.Now()
+			str = fmt.Sprintf("%s.%09d%s", ti.Format("06/02/01 15:04:05"), ti.Nanosecond(), str)
+		}
+		str = fmt.Sprintf("%-2s%s", lvlStr, str)
+
+		l.Log(lvl, str)
+
+		if l.useColors {
+			ct.ResetColor()
+		}
 	}
 }
 
-func fg(c ct.Color, bright bool) {
-	if useColors {
+func fg(l *LoggerInfo, c ct.Color, bright bool) {
```

Comment:
This will now be called by every Logger, and the corresponding `ct.ResetColor` also. Better is to put these calls into `stdLogger.Log`, then only this logger will set/reset the colors.

Reply:
The problem is that I don't have access to the `LoggerInfo` (and thus `useColors`) from the `stdLogger.Log`. Thus I cannot call `fg` from here. If we want to only have those colors for the `stdLogger`, one way would be to check that the logger we are iterating on is the first one (index 0) when we do those color changes, we already rely on this fact anyway. Or maybe there you can think of another solution?

Reply:
True - I supposed that the `LoggerInfo` _are_ the loggers, which is not the case. I can think of two ways:

1. Create a `Logger`-interface that as `SetDebugVisible`, `GetDebugVisible`... and  then it would be `NewFileLogger(name string)(int, error)`, `Logger` would be a simple `io.WriteCloser`, and you would have a `Logger.SetColor` that would be empty for most loggers.

2. Change the signature of `Logger.Log` to `Log(msg string, color ct.Color)`, because you don't need the `lvl`, but only the color.

I have a small preference for 1., but 2. seems to be OK to me, too.

Reply:
For 1., I'm not sure I understand the interface/struct logic you describe. What would be wrapping what? Where do I store the info about each logger (`debugLvl`, `showTime`, ...)? Do we have 2 interfaces, one being a `io.WriteCloser` and another one having `SetDebugVisible `, `SetUseColors`, etc... ?

For 2., how do I differentiate between cases where I have to write in stdOut or stdErr? It seems to me that I still need the `lvl` (in the case of the `stdLogger` at least). And if we do that, in order to check which `color` I should give, I'd still have to check if this is the `stdLogger` or not (checking the logger's index in `loggers`). Or maybe I misunderstood a part?

I have a 3rd alternative: in `NewFileLogger` and `NewSyslogLogger`, we explicitly put the `useColors` to `false`. In general, if you want to force a certain value for any `Logger`, you set it in the function used to create this `Logger`.

Reply:
1. each logger would store the info itself. Instead of reading out the field, `log` would have to call `Logger.GetDebugVisible`. The `Logger` interface would also implement the `io.WriteCloser`, so everybody implementing `Logger` would have to add `(Set|Get)Debug(Visible|Color|Time)`, `SetColor`, `Close` and `Write`.

2. the color would be calculated in `log` and then passed to the `Logger`. All non-std loggers would simply ignore that field. But you're right, it would be better to pass the `lvl` instead of `color`, then the std-logger could set the colors appropriately and the other loggers could do something if it is a `Warn`, `Error` or `Fatal`.

3. I don't like the fact that `log` would still be handling the colors, although they will not be used for the non-std loggers. 

My new favourite is 2. - I think it would be the least invasive for what you have now and cleanly separate the colors from the non-std loggers.

Reply:
Ok, thanks for the clarifications. About 2., how do we know if we want to use colors inside `stdLogger.Log`? We don't have access to `useColors`. So I can pass the `lvl` and compute the color it would become if we want to use it, but we cannot know if we want to use it. What I could do is indeed pass the color inside `Log`, but I should also have `lvl` as mentioned before, for the differentiation between `stdOut` and `stdErr`, and maybe for other purposes in future loggers as you said.
Another solution is to put `useColors` as an argument for `Log`, and compute the color inside `stdLogger.Log`.
What do you think?

Reply:
Or you just pass the whole `LoggerInfo` to log - that's perhaps a little bit easier than 1.

Reply:
Indeed. I've done this one, so that we can use all those information as we want in `Log`.